lib/msal-browser/src/cache/BrowserCacheManager.ts

```diff
@@ -0,0 +1,617 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Constants, PersistentCacheKeys, StringUtils, AuthorizationCodeRequest, ICrypto, AccountEntity, IdTokenEntity, AccessTokenEntity, RefreshTokenEntity, AppMetadataEntity, CacheManager, ServerTelemetryEntity, ThrottlingEntity, ProtocolUtils } from "@azure/msal-common";
+import { CacheOptions } from "../config/Configuration";
+import { CryptoOps } from "../crypto/CryptoOps";
+import { BrowserAuthError } from "../error/BrowserAuthError";
+import { BrowserCacheLocation, BrowserConstants, TemporaryCacheKeys } from "../utils/BrowserConstants";
+import { BrowserStorage } from "./BrowserStorage";
+import { MemoryStorage } from "./MemoryStorage";
+import { IWindowStorage } from "./IWindowStorage";
+
+/**
+ * This class implements the cache storage interface for MSAL through browser local or session storage.
+ * Cookies are only used if storeAuthStateInCookie is true, and are only used for
+ * parameters such as state and nonce, generally.
+ */
+export class BrowserCacheManager extends CacheManager {
+
+    // Cache configuration, either set by user or default values.
+    private cacheConfig: CacheOptions;
+    // Window storage object (either local or sessionStorage)
+    private browserStorage: IWindowStorage;
+    // Client id of application. Used in cache keys to partition cache correctly in the case of multiple instances of MSAL.
+    private clientId: string;
+    private cryptoImpl: CryptoOps;
+
+    // Cookie life calculation (hours * minutes * seconds * ms)
+    private readonly COOKIE_LIFE_MULTIPLIER = 24 * 60 * 60 * 1000;
+
+    constructor(clientId: string, cacheConfig: CacheOptions, cryptoImpl: CryptoOps) {
+        super();
+
+        this.cacheConfig = cacheConfig;
+        this.clientId = clientId;
+        this.cryptoImpl = cryptoImpl;
+
+        this.browserStorage = this.setupBrowserStorage(cacheConfig.cacheLocation);
+
+        // Migrate any cache entries from older versions of MSAL.
+        this.migrateCacheEntries();
+    }
+
+    /**
+     * Returns a window storage class implementing the IWindowStorage interface that corresponds to the configured cacheLocation.
+     * @param cacheLocation 
+     */
+    private setupBrowserStorage(cacheLocation: string): IWindowStorage {
+        if (cacheLocation !== BrowserCacheLocation.MemoryStorage) {
+            try {
+                return new BrowserStorage(cacheLocation);
+            } catch (e) {
+                this.cacheConfig.cacheLocation = BrowserCacheLocation.MemoryStorage;
+                // TODO: Log error here
+            }
+        }
+
+        return new MemoryStorage();
```

Comment:
Potential alternative:

```suggestion
        switch (cacheLocation) {
            case BrowserCacheLocation.LocalStorage:
            case BrowserCacheLocation.SessionStorage:
                try {
                    return new BrowserStorage(cacheLocation);
                } catch (e) {
                    this.cacheConfig.cacheLocation = BrowserCacheLocation.MemoryStorage;
                    return new MemoryStorage();
                }
            case BrowserCacheLocation.MemoryStorage:
                return new BrowserStorage(cacheLocation);
        }
```

Reply:
line 61 should be `return new MemoryStorage();` but I see your point. I was trying to do this with a single `if` statement, which is why it is the way it is. I can change to using switch if it's more readable.

Reply:
changed to switch statement